pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -272,17 +288,39 @@ private PersistentSubscription createPersistentSubscription(String subscriptionN
 
     @Override
     public void publishMessage(ByteBuf headersAndPayload, PublishContext publishContext) {
-        MessageDeduplication.MessageDupStatus status = messageDeduplication.isDuplicate(publishContext, headersAndPayload);
-        switch (status){
-            case NotDup:
-                ledger.asyncAddEntry(headersAndPayload, this, publishContext);
-                break;
-            case Dup:
-                // Immediately acknowledge duplicated message
-                publishContext.completed(null, -1, -1);
-                break;
-            default:
-                publishContext.completed(new MessageDeduplication.MessageDupUnknownException(), -1, -1);
+        pendingWriteOps.incrementAndGet();
+        if (isFenced) {
+            publishContext.completed(new TopicFencedException("fenced"), -1, -1);
+            decrementPendingWriteOpsAndCheck();
+            return;
+        } else {
+            MessageDeduplication.MessageDupStatus status = messageDeduplication.isDuplicate(publishContext, headersAndPayload);
+            switch (status) {
+                case NotDup:
+                    ledger.asyncAddEntry(headersAndPayload, this, publishContext);
+                    break;
+                case Dup:
+                    // Immediately acknowledge duplicated message
+                    publishContext.completed(null, -1, -1);
+                    pendingWriteOps.decrementAndGet();
+                    break;
+                default:
+                    publishContext.completed(new MessageDeduplication.MessageDupUnknownException(), -1, -1);
+                    pendingWriteOps.decrementAndGet();
+            }
+        }
+    }
+
+    private void decrementPendingWriteOpsAndCheck() {
+        long pending = pendingWriteOps.decrementAndGet();
+        if (pending == 0 && isFenced) {
+            synchronized (this) {
+                if (isFenced) {
+                    messageDeduplication.resetHighestSequenceIdPushed();
```

Comment:
Maybe we should do and info log to signal we're opening the topic again.

Reply:
will add